src/rpc/mining.cpp

```diff
@@ -123,10 +125,19 @@ UniValue generateBlocks(boost::shared_ptr<CReserveScript> coinbaseScript, int nG
           // LOCK(cs_main);
             IncrementExtraNonce(pblock, nExtraNonce);
         }
-        while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
+
+        const uint32_t req_nBits = generateWeak > 0 ? WeakBlockProofOfWork(pblock->nBits) :pblock->nBits;
+
+        do {
             ++pblock->nNonce;
-            --nMaxTries;
-        }
+            while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetHash(), req_nBits, Params().GetConsensus())) {
```

Comment:
noob nitpick: might want to pull some of these out to named variables for easier reading.